datastream-bigquery/src/main/java/com/linkedin/datastream/bigquery/BigqueryBatchCommitter.java

```diff
@@ -86,56 +104,70 @@ public BigqueryBatchCommitter(VerifiableProperties properties) {
             LOG.error("Unable to read credentials: {}", credentialsPath);
             throw new RuntimeException(e);
         }
+    }
 
-        this._numOfCommitterThreads = properties.getInt(CONFIG_THREADS, 1);
-        this._executor = Executors.newFixedThreadPool(_numOfCommitterThreads);
+    private synchronized void createOrUpdateTable(String destination) {
+        final String[] datasetTableNameRetention = destination.split("/");
 
-        this._destTableSchemas = new ConcurrentHashMap<>();
-        this._destTableCreated = new HashMap<>();
-    }
+        final Schema desiredTableSchema = _destTableSchemas.get(destination);
 
-    private synchronized void createTableIfAbsent(String destination) {
-        if (_destTableCreated.containsKey(destination)) {
-            return;
+        final long partitionRetentionDays = Long.parseLong(datasetTableNameRetention[2]);
+        final TimePartitioning timePartitioning;
+        if (partitionRetentionDays > 0) {
+            timePartitioning = TimePartitioning.of(TimePartitioning.Type.DAY, partitionRetentionDays * 86400000L);
+        } else {
+            timePartitioning = TimePartitioning.of(TimePartitioning.Type.DAY);
         }
-        String[] datasetTableNameRetention = destination.split("/");
 
-        try {
-            TableId tableId = TableId.of(datasetTableNameRetention[0], sanitizeTableName(datasetTableNameRetention[1]));
-            TableDefinition tableDefinition;
-            long partitionRetentionDays = Long.parseLong(datasetTableNameRetention[2]);
-            if (partitionRetentionDays > 0) {
-                tableDefinition = StandardTableDefinition.newBuilder()
-                        .setSchema(_destTableSchemas.get(destination))
-                        .setTimePartitioning(
-                                TimePartitioning.of(TimePartitioning.Type.DAY, partitionRetentionDays * 86400000L))
-                        .build();
+        final TableId tableId = TableId.of(datasetTableNameRetention[0], sanitizeTableName(datasetTableNameRetention[1]));
+
+        final Optional<Table> optionalExistingTable = Optional.ofNullable(_bigquery.getTable(tableId));
+        if (optionalExistingTable.isPresent()) {
+            final Table existingTable = optionalExistingTable.get();
+            final Schema existingTableSchema = Optional.ofNullable(existingTable.getDefinition().getSchema())
+                    .orElseThrow(() -> new IllegalStateException(String.format("schema not defined for table: %s", tableId)));
+            if (!desiredTableSchema.equals(existingTableSchema)) {
```

Comment:
Your approach of checking for schema changes against destination table sounds good.
But, you might need some kind of sequencing so that you don't rollback evolution in another thread handling different partition, which hasn't see evolution just yet.

Things might get consistent eventually, but there can be lot of noise in the meantime.

Reply:
Agreed. I was thinking about implementing a distributed lock using Zookeeper, but didn't get to that point just yet. In theory the chances of schema rollback/conflicting evolution should be relatively small, but the possibility is there. I'll see about adding the locking in a future commit.

Reply:
"_In theory the chances of schema rollback/conflicting evolution should be relatively small_" this may be true in some cases. But in production with some topics with about 500 partitions, the odds of this happening is higher.

The concern is more about what to expect when two committers work with different versions of the schema. How does the system react? We need to run through different schema evolution scenarios and see how severe this case is.

Suppose you have to handle this scenario for some reason. One possible way you to handle this is by keeping track of schema versions (maybe schema registry can give you that) and only evolve in one direction. Only update the schema to a higher version and never the other way. Again, it depends on schema evolution scenarios.

I will let you suggest what the right path here is.

Reply:
I thought about this a bit more over the weekend, and I think we are fairly safe with the way schema evolution is implemented with BigQuery. BigQuery only allows table schema to evolve in certain ways without requiring a table rewrite, which we are not currently planning on supporting.

- When adding a column to an existing table, the column must be optional/nullable. https://cloud.google.com/bigquery/docs/managing-table-schemas#manually_adding_an_empty_column
- A column can be soft-deleted by marking the column as optional/nullable, BigQuery does not allow hard-deleting a column. https://cloud.google.com/bigquery/docs/managing-table-schemas#relaxing_a_columns_mode

With these restrictions, it should be safe for multiple threads/nodes to add columns and soft-delete columns without having to worry about reverting the table schema to a previous state. I ran an integration test locally to confirm that BigQuery would allow me to evolve the table schema following the above rules and would throw an exception if I tried to violate them, i.e. remove a column from the schema that was added in a previous update. Based on my testing, I plan to make a small tweak to schema evolution to handle the case where there is a race condition updating the schema. The node that loses the race would eventually recover, but this small tweak should allow the schema evolution and commit to go through without having to cancel and rebuild the whole batch.

Reply:
Yep, that is what I thought. I am good as long as we have full test coverage around this case.